### ISSUE TITLE
Ganeti QA suite OS checks are build around a specific OS provider

### ISSUE THREAD

**Issue by rbott** [ganeti/ganeti#1426]
The Ganeti QA suite includes the option to use a specific OS provider. Using `noop` here works to pass the QA (sort of), but leads to the problem that there obviously is no OS installed inside the instances. That again leads to the default 120 seconds shutdown-timeout for each shutdown/failover because there is nothing inside the instance which could follow the ACPI shutdown signal. This unnecessarily slows down the whole QA process.

Currently it is not possible to just set "any" OS provider (e.g. `debootstrap`) because the QA suite seems to be hardcoded around [this one](https://github.com/osuosl/ganeti-instance-image). In several instances the QA suite passes the `--os-parameters` option with `os-image=[...]` as argument which supposedly tells `ganeti-instance-image` to download a predefined disk image via HTTP from a given URL (QA parameter `qa-storage`) which seems like a good idea from a QA point of view.

Setting the OS provider to e.g. debootstrap will not work as it stumbles over the "os-image" parameter which it does not understand/need.

I am not exactly sure how to proceed here? Should we..
* try to include `ganeti-instance-image` upstream as a default solution?
* rework the QA suite to support other OS providers (although using something as complex as debootstrap would make the outcome of a QA run less deterministic)
* write a new, very simplistic image based OS provider and include it upstream (just for QA purposes, e.g. like the fake hypervisor)

Any ideas on this?

**Comment by saschalucas** [comment]
can possibly the code be change to use `--shutdown-timeout 0`?

**Comment by saschalucas** [comment]
so here is the minimalist OS, that can power down on ACPI PWR-button. It lives entirely in the ramdisk, uses host kernel and busybox (nothing else). Start qa-instances with hypervisor parameters `kernel_path=/vmlinuz initrd_path=/tmp/qa_initrd kernel_args='init=/init'`(maybe set as default cluster parameters?)

the `/tmp/qa_initrd` is create by the following script:
```bash
#!/bin/bash

set -e -u
tmp="$(mktemp -d)"
mkdir --parents ${tmp}/{bin,dev/input,etc/acpi/PWRF,lib,lib64,proc,root,sbin,sys,usr/bin,usr/sbin}

echo '#!/bin/busybox sh
poweroff -f' > ${tmp}/etc/acpi/PWRF/00000080

chmod +x ${tmp}/etc/acpi/PWRF/00000080
cp --archive /dev/{null,console,tty} ${tmp}/dev/
cp --archive /dev/input/event* ${tmp}/dev/input
cp --archive /bin/busybox ${tmp}/bin/

echo '#!/bin/busybox sh
/bin/busybox --install 
mount -t proc none /proc
mount -t sysfs none /sys
acpid -d' > ${tmp}/init

chmod +x ${tmp}/init
cd ${tmp}
find . -print0 | cpio --null --create --verbose --format=newc | gzip --best > /tmp/qa_initrd
cd -
rm -rf ${tmp}
```

**Comment by iustin** [comment]
Originally, the QA suite was based on a different (internal, and I think never published) minimal image, and not debootstrap, because that's not stable enough; it was simply a tar dump of a pre-configured minimal debian. Ideally, the QA suite also would check instance health (since otherwise the QA run doesn't check if instances actually work), so a really minimal instance as above is a bit too minimalistic. I don't know if QA suite nowadays still has the hooks for checking instance health though.

In any case, we should have some image for QA. Starting with the one @saschalucas provided is good enough.

**Comment by rbott** [comment]
@saschalucas 
> can possibly the code be change to use `--shutdown-timeout 0`?
Currently this is not possible and QA code would have to be changed for that (possibly in a configurable way). However, hitting the shutdown-timeout always means that the instance will be in the `Error_down` state instead of `Admin_down` which is probably not desirable for a QA suite :-)

Your attempt for a diskless minimalist ACPI-understanding OS looks very promising! I will try to incorporate this into my QA setup and see if this could/should somehow be integrated upstream.

@iustin 
> Originally, the QA suite was based on a different (internal, and I think never published) minimal image,
But how was this image installed? With the OS provider I mentioned above? 
> I don't know if QA suite nowadays still has the hooks for checking instance health though.
Yes, it supports calling an external script (which receives some information like the instance name via environment vars). But said script (or at least an example/reference implementation) is not part of the QA right now.

**Comment by iustin** [comment]
@rbott:

>  But how was this image installed? With the OS provider I mentioned above?

No, not that one. We had a very small in-house image. But the image is not that important; all one needs is a reproducible image, no matter if originally created via debootstrap.

**Comment by rbott** [comment]
@iustin:
But the image itself is only half the story, right? I would still need to write my own very simple OS provider which actually writes this image to my instance disk. Am I correct here?

**Comment by iustin** [comment]
> @iustin:
> But the image itself is only half the story, right? I would still need to write my own very simple OS provider which actually writes this image to my instance disk. Am I correct here?

Ah, yes, that's indeed correct. I'll have to check if I don't (still) have that OS definition somewhere.

**Comment by iustin** [comment]
Yay, success. I found this: `ganeti-instance-etch-image_0.0.20080530-1_all.deb` somewhere in my filesystem :)

Since it's from 2008, I think it's before QA scripts got validation, but at least as minimal thing it should be a simple start.

I also found a year later one, for OS API 10, a `ganeti-instance-busybox_0.2.1-1_all.deb`.

I'll have to check to see which one is worth resurrecting, maybe none, or maybe if I find a later one in some backups.

**Comment by apoikos** [comment]
> @iustin:
> But the image itself is only half the story, right? I would still need to write my own very simple OS provider which actually writes this image to my instance disk. Am I correct here?

Not necessarily. For instance, what @saschalucas proposes requires no OS provider, just setting the `kernel_path` and `initrd_path`hypervisor parameters (which I find a brilliant idea btw).

**Comment by rbott** [comment]
I have tried that path but I could not yet get it working. I am using a slightly modified version of the above script by @saschalucas:
```
#!/bin/bash

set -e -u
tmp="$(mktemp -d)"
mkdir --parents ${tmp}/{bin,dev/input,etc/acpi/PWRF,lib,lib64,proc,root,sbin,sys,usr/bin,usr/sbin,modules}

echo '#!/bin/busybox sh
poweroff -f' > ${tmp}/etc/acpi/PWRF/00000080

chmod +x ${tmp}/etc/acpi/PWRF/00000080
cp --archive /dev/{null,console,tty} ${tmp}/dev/
cp --archive /dev/input/event* ${tmp}/dev/input
cp --archive /bin/busybox ${tmp}/bin/

for MODULE in $(find /lib/modules -name 'button.ko'); do
	cp $MODULE ${tmp}/modules/;
done

echo '#!/bin/busybox sh
/bin/busybox --install 
mount -t proc none /proc
mount -t sysfs none /sys
insmod /modules/button.ko
acpid -d' > ${tmp}/init

chmod +x ${tmp}/init
cd ${tmp}
find . -print0 | cpio --null --create --verbose --format=newc | gzip --best > /boot/ganeti_busybox_initrd
cd -
rm -rf ${tmp}
```
Due to the following error message I tried adding the ACPI "button" module to the initramfs (it is not included in the default Debian Buster kernel). However, the busybox acpid always complains about /dev/input/event0, with or without the module:
```
[    3.528903] Run /init as init process
[    3.541100] input: Power Button as /devices/LNXSYSTM:00/LNXPWRBN:00/input/input2
[    3.555161] ACPI: Power Button [PWRF]
acpid: /dev/input/event0: No such device or address
[    3.572561] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
```
Not creating /dev/input/event0 will result in a "No such file or directory" error from acpid. Does anyone have any ideas here? My related ganeti KVM settings are as follows:
```
    initrd_path: /boot/ganeti_busybox_initrd
    kernel_args: init=/init
    kernel_path: /vmlinuz
    root_path: /dev/vda1
    serial_console: True
```

For reference I have attached the entire boot log 
[console.log](https://github.com/ganeti/ganeti/files/3992972/console.log)

**Comment by iustin** [comment]
If the used kernel is not monolithic, then you need in the image all the modules as well?

**Comment by apoikos** [comment]
> I have tried that path but I could not yet get it working. I am using a slightly modified version of the above script by @saschalucas:
> 
> ```
> #!/bin/bash
> 
> set -e -u
> tmp="$(mktemp -d)"
> mkdir --parents ${tmp}/{bin,dev/input,etc/acpi/PWRF,lib,lib64,proc,root,sbin,sys,usr/bin,usr/sbin,modules}
> 
> echo '#!/bin/busybox sh
> poweroff -f' > ${tmp}/etc/acpi/PWRF/00000080
> 
> chmod +x ${tmp}/etc/acpi/PWRF/00000080
> cp --archive /dev/{null,console,tty} ${tmp}/dev/
> cp --archive /dev/input/event* ${tmp}/dev/input
> cp --archive /bin/busybox ${tmp}/bin/
> 
> for MODULE in $(find /lib/modules -name 'button.ko'); do
> 	cp $MODULE ${tmp}/modules/;
> done
> 
> echo '#!/bin/busybox sh
> /bin/busybox --install 
> mount -t proc none /proc
> mount -t sysfs none /sys
> insmod /modules/button.ko
> acpid -d' > ${tmp}/init

You probably need either `acpid -d -f`, or some kind of sleep loop (e.g. `while 1; do sleep 300; done`) otherwise `init` will exit prematurely.

> 
> chmod +x ${tmp}/init
> cd ${tmp}
> find . -print0 | cpio --null --create --verbose --format=newc | gzip --best > /boot/ganeti_busybox_initrd
> cd -
> rm -rf ${tmp}
> ```
> 
> Due to the following error message I tried adding the ACPI "button" module to the initramfs (it is not included in the default Debian Buster kernel). However, the busybox acpid always complains about /dev/input/event0, with or without the module:
> 
> ```
> [    3.528903] Run /init as init process
> [    3.541100] input: Power Button as /devices/LNXSYSTM:00/LNXPWRBN:00/input/input2
> [    3.555161] ACPI: Power Button [PWRF]
> acpid: /dev/input/event0: No such device or address
> [    3.572561] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
> ```
> 
> Not creating /dev/input/event0 will result in a "No such file or directory" error from acpid. Does anyone have any ideas here?

Yes, you probably need to `mknod` it, as your initrd does not have udevd installed.

**Comment by rbott** [comment]
Regarding acpid:
It runs with the -d flag which includes the -f - the process runs in foreground (I have double checked this already).

Regarding mknod:
The script does not use mknod, but it sets up the correct device nodes:
```
cp --archive /dev/{null,console,tty} ${tmp}/dev/
cp --archive /dev/input/event* ${tmp}/dev/input
cp --archive /bin/busybox ${tmp}/bin/
```
This is probably not a "by-the-book-way", but it in my case it leads to /dev/input/event[0-4] with the correct majors/minors.

**Comment by apoikos** [comment]
@rbott, ah, I assumed `-d` did not imply `-f`, my bad. I also just noticed you copy the device nodes. I'll play around a bit and see if I can get it to work.

**Comment by apoikos** [comment]
@rbott: apparently we also need `evdev.ko` :)

**Comment by rbott** [comment]
@apoikos: dang, now it works, thanks! :man_facepalming: 

I will include this in my QA setup and see what happens. After activating tests for DRBD, file, sharedfile and plain the QA takes _very_ long now (mostly because of many many many 120 seconds shutdown timeouts). Let's see how this improves the situation :-)

**Comment by rbott** [comment]
That was very successful, the QA runtime dropped down from over 2 hours to roughly 1 hour.

The question is, how to include this in the QA suite in a helpful way? Setting the correct KVM parameters is already possible through `hypervisor-parameters` in the json config file. However, we either need to ship a custom built kernel + initramfs with the QA (and have the QA suite install it on all nodes) or have the above script shipped with the QA (and have the QA suite install & run that on all nodes).

Any suggestions? I would propose a solution like this:
* add a QA parameter to enable a minimalist OS for instance tests (which would then break/disable OS providers, although as noted above, the QA suite really just supports one specific OS provider right now)
* add a QA parameter to set the path of a linux kernel image (which needs to be provided by the user and could/should already include the initramfs)
* have the QA suite distribute the kernel image to all nodes and set KVM parameters accordingly (this should in theory also work for XEN)
